src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/MsBuildModelWatcher.cs

```diff
@@ -0,0 +1,63 @@
+﻿using Microsoft.Build.Evaluation;
+using Microsoft.VisualStudio.ProjectSystem.Utilities;
+using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+using System.Threading.Tasks;
+using System.Threading;
+using System.ComponentModel.Composition;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Editor
+{
+    [Export(typeof(IMsBuildModelWatcher))]
+    internal class MsBuildModelWatcher : OnceInitializedOnceDisposedAsync, IMsBuildModelWatcher
+    {
+        private readonly IProjectThreadingService _threadingService;
+        private readonly IFileSystem _fileSystem;
+        private readonly IMsBuildAccessor _accessor;
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private string _tempFile;
+
+        [ImportingConstructor]
+        public MsBuildModelWatcher(IProjectThreadingService threadingService, 
+            IFileSystem fileSystem, 
+            IMsBuildAccessor accessor, 
+            UnconfiguredProject unconfiguredProject) :
+            base(threadingService.JoinableTaskContext)
+        {
+            _threadingService = threadingService;
+            _fileSystem = fileSystem;
+            _accessor = accessor;
+            _unconfiguredProject = unconfiguredProject;
+        }
+
+        public void Initialize(string tempFile)
+        {
+            _tempFile = tempFile;
+            _threadingService.ExecuteSynchronously(() => InitializeAsync());
+        }
+
+        public void ProjectXmlHandler(object sender, ProjectXmlChangedEventArgs args)
```

Comment:
Hi, Fred, I don't think we should hook up the msbuild collection event here.  The ProjectLockService actually fire Releasing/ReleasedWriteLock event, which contains the exact set of project files are changed, and renamed.  That is exactly what you want, but those two events are internal now.  I think we should make them public, and you change should chain to that.  You don't want to write file for every xml change, because it will be a performance problem...